### DISCUSSION TITLE
ERROR ITMS-4088 when uploading to TestFlight (iOS)

### DISCUSSION THREAD

**Discussion by alkla** [fastlane/fastlane#17684] [Q&A]
Hi!

I'm using Fastlane for Beta Upload to Apple TestFlight and get the following error, that I can't understand

```
ERROR ITMS-4088: "The organization [...] does not have permission to modify the application with the Apple ID [Id]. Please try again, using an account for an organization with access to this application." at SoftwareAssets
Return status of iTunes Transporter was 1: ERROR ITMS-4088: "The organization [...] does not have permission to modify the application with the Apple ID [Id]. Please try again, using an account for an organization with access to this application." at SoftwareAssets
The call to the iTMSTransporter completed with a non-zero exit status: 1. This indicates a failure.
```
I checked the Apple ID and Team ID and they were both correct. I got the team id from this json: https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/ra/user/detail

Upload to TestFlight via Xcode and other 3rd party tools works perfectly.

It is maybe worth mentioning that I'm a member of multiple teams in AppStore Connect.

What am I doing wrong?

Xcode 12.2
Fastlane version 2.168.0

**Comment by ainame** [comment]
Your account need to be at least "App Manager" role for `pilot`. Have you checked that?
https://docs.fastlane.tools/actions/testflight/#role-for-app-store-connect-user

See also this. "Developer" has only read-only access for managing TestFlight builds.

<img width="828" alt="Screenshot 2020-11-29 at 04 32 36" src="https://user-images.githubusercontent.com/748949/100533431-07f5c900-31fc-11eb-89b3-035f1007fb9f.png">

https://developer.apple.com/support/roles/

**Comment by alkla** [comment]
@ainame Thanks for taking the time to respond!
My account is both Account Holder and admin, so that shouldn't be the problem.

Uploading works when I use Xcode or external tools to upload, so I don't really understand why this is happening.

**Comment by ainame** [comment]
Hmm that's strange 🤔 
I feel like you should report this as a bug with more details in issues so that it can catch more eyes from contributors?

**Comment by khalilelhajji** [comment]
@alkla i have the same issue !
I'm interested if you found a solution or a workaround :)

**Comment by alkla** [comment]
@khalilelhajji No, I just use a third party GUI tool for uploading, a bit frustrating...

**Comment by rogerluan** [comment]
@khalilelhajji @alkla 

Have you tried playing around with the team_id, team_name and dev_portal_team_id? They're described [here](https://docs.fastlane.tools/actions/testflight/):

`team_id` - The ID of your App Store Connect team if you're in multiple teams
`team_name`- The name of your App Store Connect team if you're in multiple teams
`dev_portal_team_id` - The short ID of your team in the developer portal, if you're in multiple teams. Different from your iTC team ID!

I honestly don't know how to get each of them exactly, but one of them look like `123456789`, the other looks like `RF1122AB99` 😅

**Comment by absoftware** [comment]
I have a problem that when I build 3 applications one after another then I can build only apps for one AppStore Team. If script goes to the app from another team then it fails and I get for second upload (Team 2) following error:

```
[Transporter Error Output]: ERROR ITMS-4088: The organization Team1 AS does not have permission to modify the application with the Apple ID 0123456789. Please try again, using an account for an organization with access to this application. at SoftwareAssets
```

as it wants to use first AppStore Team for all next builds. Do you have solution for that? How to force switching to another team before making next build?

At the beginning I was thinking that I have wrong order and I was doing:
* `build_app` for team 1
* `build_app` for team 2
* `build_app` for team 3
* `upload_to_testflight` for team 1
* `upload_to_testflight` for team 2 -> error occurs here
* `upload_to_testflight` for team 3

So I changed it to
* `build_app` and `upload_to_testflight` for team 1
* `build_app` and `upload_to_testflight` for team 2 -> error occurs here
* `build_app` and `upload_to_testflight` for team 3

Of course I specify all possible parameters for these functions that it could identify build, project and team. But it’s doesn’t work. It works until I make a builds only for one team per script. (edited: I had missing `itc_provider` param. See comments below.)

PS. In all teams I’m at least App Manager or Admin. So it’s not a case with missing access. Especially that error says about Team 1 when I do upload for Team 2.

Should I report it as new issue or we can take an action here?

**Reply by lucgrabowski** [reply]
>I specify all possible parameters for these functions that it could identify build, project and team

Which parameters do you specify? `team_id`, `team_name` as @rogerluan  advised? If that's not working, you probably should also add `itc_provider`. You could check it with `iTMSTransporter` or install Apple's Transporter app, upload a build and look for the value in logs.

**Reply by absoftware** [reply]
@lucgrabowski Thank you for the correct answer. I spent entire night searching this. And indeed I had missing param `itc_provider`. This one was not so obvious because of missing word `team` in name of this param.

All possible `itc_provider` values can be listed using command
```
/Applications/Xcode.app/Contents/SharedFrameworks/ContentDeliveryServices.framework/Versions/A/itms/bin/iTMSTransporter  -m provider -u '<my-apple-id>' -p '<app-specific-password>' -account_type itunes_connect -v off
```

So finally I ended with
```
upload_to_testflight(
    team_id: config["team"]["team_id"], # only digits
    team_name: config["team"]["team_name"], # full name with spaces
    dev_portal_team_id: config["team"]["dev_portal_team_id"], # letters and digits
    itc_provider: config["team"]["itc"], # short name without spaces
    ipa: "#{config["target"]}.ipa",

    # These options are required by env FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD 
    # otherwise it will ask about username/password. The correct value should be taken from Apple ID
    # property in the App Information section in App Store Connect. Define in your shell:
    # export FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD="your-app-id-pass"
    username: config["app_store_username"],
    apple_id: config["app_store_app_id"],
    skip_waiting_for_build_processing: false,
    reject_build_waiting_for_review: true,

    groups: config["tester_groups"],

    localized_build_info: {
        "default": {
            whats_new: changelog_text,
        }
    }
)
```

So I use now 4 parameters to identify the team. Something went wrong with that implementation but I suspect it's historical issue coming from merging Developer Program with iTunes Connect.

**Reply by lucgrabowski** [reply]
I'm glad it worked :)

**Comment by lucgrabowski** [comment]
Hi @alkla, have you tried adding `itc_provider` parameter?
